woocommerce/woocommerce
# [Settings] Handle outputed script tags

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/52975

This PR extracts `<script>` tag contents from custom settings fields using a DOM parser and injects them into the DOM.

Additionally, this PR fixes the type definitions:

1. Update `type` of `SettingsField` to be explicitly defined constants instead of `string`.
2. Add `CustomSettingsField` and fix the `GroupSettingsField` type.
3. Fix SettingsSection.settings type. (`SettingField` is undefined.)

I use the same way as POC https://github.com/woocommerce/woocommerce/pull/48340 to inject the script into the DOM. Since it's usually dangerous to do this, I think we should look at it more deeply and ensure that injecting the script into the DOM is safe. Also, we need to clean up the script when the component unmounts. However, if scripts create global variables or event listeners, it might cause unexpected behavior and cause memory leaks. But this is beyond the scope of this PR. 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Add/Enable Gutenberg version `19.8.0` or earlier.
2. Turn on the `settings` feature flag
3. Use the [Settings Tester](https://github.com/woocommerce/woocommerce-settings-migration-tester) to test Modern screens. Clone it locally and `npm install && npm run plugin-zip` / [settings-tester.zip](https://github.com/user-attachments/files/18208344/settings-tester.zip)
3. Upload and activate the Tester
4. Go to `/wp-admin/admin.php?page=wc-settings&tab=settings-tester&section=javascript`
5. Toggle `Echoing HTML` checkbox and confirm the yellow background is shown when the checkbox is checked.


https://github.com/user-attachments/assets/b9d31e78-c1b4-447b-8077-138cf531c205




### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

This is behind the settings feature flag.

</details>
